languages/python/exercises/concept/dicts/.meta/design.md

```diff
@@ -0,0 +1,59 @@
+# Design
+
+## Goal
+
+The goal of this exercise is to teach the basics of the `dict` (dictionary, mapping) data type in Python.
+
+## Things to teach
+
+- create a `dict` via constructor and literal.
+- access values in a `dict` via `keys`.
+- assign values in a `dict` via key reference.
+- check for membership of a key in a given dictionary.
+- add new `key`:`value` pairs to the `dict`.
+- remove `key`:`value` pairs from the `dict`.
+- iterate through a `dict` using `dict.keys()`, `dict.values()`, or `dict.items()`.
+- `dict` method `setdefault()`
+
+## Things not to teach
+
+- Dictionary comprehensions
+- `dict` methods such as `get()` or `clear()`.
+- Built-in functions as they relate to this data structure (_e.g._ `copy()`, `len()`, or `enumerate()`.
+- Sorting by `keys` or `values`.
+- Swapping `keys` for `values`.
+- Knowing that Dictionaries can be _nested_, _-- e.g._ ' a dictionary of dictionaries'.
+- Mutability
+- `copy()` vs `deepcopy()`
+- Related `collections` module with `Counter()` and `defaultdict()`
+- Memory and performance characteristics.
+
+## Concepts
+
+- `dicts`
+
+## Prerequisites
+
+- `booleans`
+- `for-loops`
+- `functions`
```

Comment:
Either we need to decide or it's been decided already and I'm out of the loop: what is the definition of the `functions` concept? Of the following, which are we going with?
1. "I'm using a function, so I need to know what functions are"
2. "I'm writing/changing a function, so I need to know what functions are"

@ErikSchierboom @BethanyG

Reply:
I would say both - not either or.  But the assumption with all of the concept exercises is that a student already knows how to program -- just with a different programming language than Python.  

Python has yet to define a `basics` exercise, so I think that's at the root of our confusion (_certainly mine, since I don't think our maintainers group had that convo explicitly_).  Here are some examples from C#, Ruby, and JavaScrpit:

-   [C# Basics](https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/basics/.docs/introduction.md)
-  [Ruby Basics](https://github.com/exercism/v3/blob/master/languages/ruby/exercises/concept/basics/.docs/introduction.md)
-  [JavaScript Basics](https://github.com/exercism/v3/blob/master/languages/javascript/exercises/concept/basics/.docs/introduction.md)

Once we have a `basics`, that will be the pre-requisite here.  I can work writing up a `basics` concept exercise issue, and start on the exercise itself.  I also have on my list to re-work both the concepts list and the table.  It feels like we might have some things at the wrong granularity level there.  Maybe we can all go back over it and discuss?

Reply:
> Python has yet to define a basics exercise, so I think that's at the root of our confusion (certainly mine, since I don't think our maintainers group had that convo explicitly).

That would do it. I've been working off the assumption that any exercise without prerequisites is considered a "starting point", and marking `functions` as a prequisite seemed like a good way to get a circular dependency. If we have support for a "starting" `basics` exercise, that works.

Reply:
I'd suggest writing up a few things for the `basics` exercise and to put that in the maintaining-python channel. We'll discuss it from there then. We can get some advice from other people that way too. What do you think?

Reply:
> @cmccandless -- I would say both - not either or. But the assumption with all of the concept exercises is that a student already knows how to program -- just with a different programming language than Python.

Yep, this ☝️ 

> Once we have a basics, that would be the pre-requisite here - and not functions. In fact, I'd contend functions as it's own exercise is probably off the table, unless we want a more advanced one that is for higher-order functions or other beyond-basic concept.

This is how existing tracks approach this. The `basics` exercise introduces students to some of the basic concepts: how to declare variables, (re-)assign values to variables, declaring functions (with parameters) and calling functions. For those tracks that require functionality to be defined in a class, they'll also include a minimal example of the class syntax and how to instantiate a class and call one of its methods.

> That would do it. I've been working off the assumption that any exercise without prerequisites is considered a "starting point", and marking functions as a prequisite seemed like a good way to get a circular dependency. If we have support for a "starting" basics exercise, that works.

Excellent.